experiments-core/src/foedus/ycsb/ycsb_driver.cpp

```diff
@@ -0,0 +1,374 @@
+DEFINE_int32(reducer_buffer_size, 1, "Size of reducer's buffer per NUMA node in GB.");
+DEFINE_int32(loggers_per_node, 1, "Number of log writers per numa node.");
+DEFINE_int32(thread_per_node, 0, "Number of threads per NUMA node. 0 uses logical count");
+DEFINE_int32(numa_nodes, 0, "Number of NUMA nodes. 0 uses physical count");
+DEFINE_bool(use_numa_alloc, true, "Whether to use ::numa_alloc_interleaved()/::numa_alloc_onnode()"
+  " to allocate memories. If false, we use usual posix_memalign() instead");
+DEFINE_bool(interleave_numa_alloc, false, "Whether to use ::numa_alloc_interleaved()"
+  " instead of ::numa_alloc_onnode()");
+DEFINE_bool(mmap_hugepages, false, "Whether to use mmap for 1GB hugepages."
+  " This requies special setup written in the readme.");
+DEFINE_int32(log_buffer_mb, 512, "Size in MB of log buffer for each thread");
+DEFINE_bool(null_log_device, false, "Whether to disable log writing.");
+DEFINE_int64(duration_micro, 1000000, "Duration of benchmark in microseconds.");
+
+// YCSB-specific options
+DEFINE_string(workload, "A", "YCSB workload; choose A/B/C/D/E.");
+DEFINE_int64(max_scan_length, 1000, "Maximum number of records to scan.");
+DEFINE_bool(read_all_fields, true, "Read all or only one field(s) in read transactions.");
+DEFINE_bool(write_all_fields, true, "Write all or only one field(s) in update transactions.");
+
+// If this is enabled, the original YCSB implementation gives a fully ordered key across all
+// threads. But that's hard to scale in high core counts. So we use [worker_id | local_count].
+DEFINE_bool(ordered_inserts, false, "Whether to make the keys ordered, i.e., don't hash(keynum).");
+
+YcsbWorkload YcsbWorkloadA('A', 0,  50U,  100U, 0);     // Workload A - 50% read, 50% update
+YcsbWorkload YcsbWorkloadB('B', 0,  95U,  100U, 0);     // Workload B - 95% read, 5% update
+YcsbWorkload YcsbWorkloadC('C', 0,  100U, 0,    0);     // Workload C - 100% read
+YcsbWorkload YcsbWorkloadD('D', 5,  100U, 0,    0);     // Workload D - 95% read, 5% insert
+YcsbWorkload YcsbWorkloadE('E', 5U, 0,    0,    100U);  // Workload E - 5% insert, 95% scan
+
+int64_t max_scan_length() {
+  return FLAGS_max_scan_length;
+}
+
+YcsbRecord::YcsbRecord(char value) {
+  // So just write some arbitrary characters provided, no need to use rnd
+  memset(data_, value, kFields * kFieldLength * sizeof(char));
+}
+
+// TODO(tzwang): make this field content random
+void YcsbRecord::initialize_field(char *field) {
+  memset(field, 'a', kFieldLength);
+}
+
+YcsbKey& YcsbKey::next(uint32_t worker_id, uint32_t* local_key_counter) {
+  auto low = ++(*local_key_counter);
+  return build(worker_id, low);
+}
+
+YcsbKey& YcsbKey::build(uint32_t high_bits, uint32_t low_bits) {
+  uint64_t keynum = ((uint64_t)high_bits << 32) | low_bits;
+  if (!FLAGS_ordered_inserts) {
+    keynum = (uint64_t)foedus::storage::hash::hashinate(&keynum, sizeof(keynum));
+  }
+  data_ = kKeyPrefix;
+  data_.append(std::to_string(keynum));
```

Comment:
heh didn't find obvious evidence on if it's heap or stack (some book says it's ostringstream behind it), i think let's use std::itoa().